python/nav/ipdevpoll/shadows/__init__.py

```diff
@@ -260,6 +267,56 @@ def _fix_binary_garbage(self):
                 setattr(self, attr, repr(value))
         self.clear_cached_objects()
 
+    @classmethod
+    def prepare_for_save(cls, containers):
+        cls.changed_devices = cls._detect_version_changes(containers)
+        return super(Device, cls).prepare_for_save(containers)
+
+    @classmethod
+    def _detect_version_changes(cls, containers):
+        """
+        Detects if the software, hardware or firmware version changed for each device.
+
+        Returns a dictionary of the device and which of the versions changed.
+        """
+        if cls not in containers:
+            return
+
+        changed_devices = {}
+        for device in containers[cls].values():
+            old_device = device.get_existing_model()
+            if old_device:
+                changed_versions = list(
+                    set(device.get_diff_attrs(old_device)).intersection(
+                        (
+                            'hardware_version',
+                            'software_version',
+                            'firmware_version',
+                        )
+                    )
+                )
+                if changed_versions:
+                    changed_devices[device] = changed_versions
+        return changed_devices
+
+    @classmethod
+    def cleanup_after_save(cls, containers):
+        cls._post_events_version_changes(containers)
+        return super(Device, cls).cleanup_after_save(containers)
+
+    @classmethod
+    def _post_events_version_changes(cls, containers):
+        """Posts events for software, hardware or firmware changes."""
+        if cls.changed_devices:
+            for shadow_device, changed_versions in cls.changed_devices.items():
+                device = shadow_device.get_existing_model()
+                for version in changed_versions:
+                    cls.event.start(
```

Comment:
This event seems stateless in nature to me, so using event.notify might be more natural here (there is a bug in that function, i have a fix in #2414)

Reply:
Not sure if this is relevant for this type of alert, but it might be good to add alertmsg templates. These (I think) are used to generate e-mail and sms alerts which might not be very interesting for this event, but the email and sms messages generated from the template is also displayed in the event history that is available somewhere in the frontend

Reply:
also might have messed up the github review thing, pretty sure this isnt how it looked when morten left review on my PRs 😥